config/settings.txt
[General]
output_dir_prefix=sim_output
verbose=false

[alpide]
chip_continuous_mode=false
data_long_enable=true
dtu_delay=10
matrix_readout_speed_fast=true
minimum_busy_cycles=8
pixel_shaping_active_time_ns=5000
pixel_shaping_dead_time_ns=200
strobe_extension_enable=false

[data_output]
data_rate_interval_ns=100000
write_event_csv=true
write_vcd=false
write_vcd_clock=false

[event]
average_event_rate_ns=2000
monte_carlo_file_type=root
qed_noise_event_rate_ns=10000
qed_noise_feed_rate_ns=5000
qed_noise_input=false
qed_noise_path=config/monte_carlo_events/pp
qed_noise_rate_ns=250
random_cluster_generation=true
random_cluster_size_mean=4
random_cluster_size_stddev=1
random_hit_generation=true
strobe_active_length_ns=100
strobe_inactive_length_ns=100
trigger_delay_ns=930
trigger_filter_enable=true
trigger_filter_time_ns=1230

[focal]
monte_carlo_file_path=config/monte_carlo_events/focal/pixel_event_tree_pythia_MB_r4cm.root
staves_per_quadrant=33

[its]
bunch_crossing_rate_ns=25
hit_density_layer0=18.6
hit_density_layer1=12.2
hit_density_layer2=9.1
hit_density_layer3=2.8
hit_density_layer4=2.7
hit_density_layer5=2.6
hit_density_layer6=2.6
hit_multiplicity_distribution_file=config/multipl_dist_raw_bins.txt
layer0_num_staves=1
layer1_num_staves=1
layer2_num_staves=1
layer3_num_staves=1
layer4_num_staves=0
layer5_num_staves=0
layer6_num_staves=0
monte_carlo_dir_path=config/monte_carlo_events/PbPb

[pct]
beam_end_coord_x_mm=275
beam_end_coord_y_mm=20
beam_start_coord_x_mm=-5
beam_start_coord_y_mm=-5
beam_step_mm=3.0
beam_time_per_step_us=125.0
layers="0;5;10;15;20;25;30;35;40"
monte_carlo_file_path=config/monte_carlo_events/focal/pixel_event_tree_pythia_MB_r4cm.root
num_layers=2
num_staves_per_layer=12
random_beam_stddev_mm=3.0
random_particles_per_s_mean=1E9
random_particles_per_s_stddev=0.5E9
time_frame_length_ns=10000

[simulation]
n_events=100
random_seed=1337
single_chip=false
system_continuous_mode=true
system_continuous_period_ns=10000
type=its


doc/pages/sim_settings.md
# Simulation Settings

The simulation model has a number of configurable settings, which the program expects to find in a ini-format settings.txt file in the directory it is executed from.
- If the settings.txt file does not exist, the program will automatically generate a new settings.txt file with the default values.
- When running a simulation, the simulation program will store a copy of settings.txt in the simulation run's data output directory, so that one can always know what settings were used for a given simulation.


## List of simulation settings

| Group       | Setting                            | Default value             | Comment                                                                                                                                                                          |
|-------------|------------------------------------|---------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| alpide      | pixel_shaping_active_time_ns       | 6000                      | Equivalent to Time over Threshold (ToT)                                                                                                                                          |
| alpide      | pixel_shaping_active_time_ns       | 200                       | Equivalent to rise time before ToT                                                                                                                                               |
| alpide      | clustering_enable                  | true                      | Enable clustering of adjacent pixel hits (and use of DATA LONG)                                                                                                                  |
| alpide      | region_fifo_size                   | 128                       | Depth of Region Readout Unit (RRU) FIFOs                                                                                                                                         |
| alpide      | region_fifo                        | 32                        | Region size. Not really used, hardcoded at the moment..                                                                                                                          |
| alpide      | matrix_readout_speed_fast          | true                      | Matrix priority encoder readout clock speed. True = 20MHz, false = 10MHz.                                                                                                        |
| alpide      | dmu_fifo_size                      | 64                        | Size of Data Management Unit (DMU) FIFO (the output "bottleneck" FIFO)                                                                                                           |
| alpide      | dtu_delay                          | 10                        | Delay (in clock cycles) to simulate delay introduced by serializing and encoding in DTU.                                                                                         |
| data_output | write_event_csv                    | true                      | Enable writing of event data (delta_t and multiplicity) to CSV file                                                                                                              |
| data_output | write_vcd                          | false                     | Enable writing SystemC signals to Value Change Dump(VCD) file (requires lots of disk space for many events)                                                                      |
| data_output | write_vcd_clock                    | false                     | Enable writing clock to VCD file (requires even more disk space)                                                                                                                 |
| simulation  | continuous_mode                    | false                     | Enable continuous mode (triggered if set to false)                                                                                                                               |
| simulation  | n_chips                            | 1                         | Number of chips to include in simulation                                                                                                                                         |
| simulation  | n_events                           | 10000                     | Number of (trigger/continuous) events to simulate                                                                                                                                |
| simulation  | random_seed                        | 0                         | Random seed. Setting to 0 will initialize random generatorswith a high entropy random seed.                                                                                      |
| event       | average_event_rate_ns              | 2500                      | Average event rate in nanoseconds                                                                                                                                                |
| event       | bunch_crossing_rate_ns             | 25                        | Bunch crossing rate/period in nanoseconds                                                                                                                                        |
| event       | hit_density_min_bias_per_cm2       | 19                        | Minimum bias hit density (traces) per square centimeter                                                                                                                          |
| event       | hit_multiplicity_distribution_file | multipl_dist_raw_bins.txt | Name/path of text file with discrete multiplicity distribution (used ifhit_multiplicity_distribution_type is set to discrete)                                                    |
| event       | hit_multiplicity_distribution_type | discrete                  | Choice of hit multiplicity distribution, either discrete or gaussian. Hitmultiplicity distribution is scaled to match the combination hit densityand number of chips.            |
| event       | hit_multiplicity_gauss_avg         | 2000                      | Average hit multiplicity in gaussian distribution (only used ifhit_multiplicity_distribution_type is set to gaussian). Not really used, and not fully implemented at the moment. |
| event       | hit_multiplicity_gauss_stddev      | 350                       | Standard deviation for hit multiplicity in gaussian distribution (only usedif hit_multiplicity_distribution_type is set to gaussian)                                             |
| event       | strobe_active_length_ns            | 4800                      | Strobe active time in nanoseconds                                                                                                                                                |
| event       | strobe_inactive_length_ns          | 200                       | Strobe inactive time in nanoseconds                                                                                                                                              |
| event       | trigger_delay_ns                   | 1000                      | Total trigger delay in nanoseconds                                                                                                                                               |
| event       | trigger_filter_enable              | true                      | Trigger filtering enable/disable (triggered mode only)                                                                                                                           |
| event       | trigger_filter_time_ns             | 10000                     | Trigger filter time in nanoseconds. If filtering is enabled, and twotriggers fall within this filter time, the last trigger(s) will be filtered(removed).                        |


src/Settings/Settings.hpp
/**
 * @file   Settings.hpp
 * @author Simon Voigt Nesbo <svn@hib.no>
 * @date   November 3, 2016
 * @brief  Header file for simulation settings file handling.
 *
 *         This file has definitions for default simulation settings, which can
 *         be used as default values or for generating the settings file if it is missing.
 */


///@defgroup settings Settings
///@{
#ifndef _SETTINGS_HPP
#define _SETTINGS_HPP

#include <QtCore>
#include <QSettings>
#include <QDir>
#include <QString>


#define DEFAULT_DATA_OUTPUT_WRITE_VCD "false"
#define DEFAULT_DATA_OUTPUT_WRITE_VCD_CLOCK "false"
#define DEFAULT_DATA_OUTPUT_WRITE_EVENT_CSV "true"
#define DEFAULT_DATA_OUTPUT_DATA_RATE_INTERVAL_NS "10000"

#define DEFAULT_SIMULATION_TYPE "its"
#define DEFAULT_SIMULATION_SINGLE_CHIP "true"
#define DEFAULT_SIMULATION_N_EVENTS "10000"
#define DEFAULT_SIMULATION_SYSTEM_CONTINUOUS_MODE "false"
#define DEFAULT_SIMULATION_SYSTEM_CONTINUOUS_PERIOD_NS "5000"
#define DEFAULT_SIMULATION_RANDOM_SEED "0"

#define DEFAULT_ALPIDE_DATA_LONG_ENABLE "true"
#define DEFAULT_ALPIDE_DTU_DELAY "10"
#define DEFAULT_ALPIDE_PIXEL_SHAPING_DEAD_TIME_NS "200"
#define DEFAULT_ALPIDE_PIXEL_SHAPING_ACTIVE_TIME_NS "6000"
#define DEFAULT_ALPIDE_MATRIX_READOUT_SPEED_FAST "true"
#define DEFAULT_ALPIDE_STROBE_EXTENSION_ENABLE "false"
#define DEFAULT_ALPIDE_MINIMUM_BUSY_CYCLES "8"
#define DEFAULT_ALPIDE_CHIP_CONTINUOUS_MODE "false"

#define DEFAULT_ITS_LAYER0_NUM_STAVES "12"
#define DEFAULT_ITS_LAYER1_NUM_STAVES "16"
#define DEFAULT_ITS_LAYER2_NUM_STAVES "20"
#define DEFAULT_ITS_LAYER3_NUM_STAVES "0"
#define DEFAULT_ITS_LAYER4_NUM_STAVES "0"
#define DEFAULT_ITS_LAYER5_NUM_STAVES "0"
#define DEFAULT_ITS_LAYER6_NUM_STAVES "0"
#define DEFAULT_ITS_HIT_MULTIPLICITY_DISTRIBUTION_FILE "config/multipl_dist_raw_bins.txt"
#define DEFAULT_ITS_BUNCH_CROSSING_RATE_NS "25"
#define DEFAULT_ITS_MONTE_CARLO_DIR_PATH "config/monte_carlo_events/PbPb"
#define DEFAULT_ITS_HIT_DENSITY_LAYER0 "18.6"
#define DEFAULT_ITS_HIT_DENSITY_LAYER1 "12.2"
#define DEFAULT_ITS_HIT_DENSITY_LAYER2 "9.1"
#define DEFAULT_ITS_HIT_DENSITY_LAYER3 "2.8"
#define DEFAULT_ITS_HIT_DENSITY_LAYER4 "2.7"
#define DEFAULT_ITS_HIT_DENSITY_LAYER5 "2.6"
#define DEFAULT_ITS_HIT_DENSITY_LAYER6 "2.6"

#define DEFAULT_PCT_LAYERS "0;5;10;15;20;25;30;35;40"
#define DEFAULT_PCT_NUM_STAVES_PER_LAYER "1"
#define DEFAULT_PCT_MONTE_CARLO_FILE_PATH "config/monte_carlo_events/pct/pct_data.xml"
#define DEFAULT_PCT_TIME_FRAME_LENGTH_NS "200"
#define DEFAULT_PCT_RANDOM_PARTICLES_PER_S_MEAN "1E9"
#define DEFAULT_PCT_RANDOM_PARTICLES_PER_S_STDDEV "0.2E9"
#define DEFAULT_PCT_RANDOM_BEAM_STDDEV_MM "0.1"
#define DEFAULT_PCT_BEAM_START_COORD_X_MM "-0.5"
#define DEFAULT_PCT_BEAM_START_COORD_Y_MM "-0.5"
#define DEFAULT_PCT_BEAM_END_COORD_X_MM "27.5"
#define DEFAULT_PCT_BEAM_END_COORD_Y_MM "2.0"
#define DEFAULT_PCT_BEAM_STEP_MM "3.0"
#define DEFAULT_PCT_BEAM_TIME_PER_STEP_US "125"

#define DEFAULT_FOCAL_MONTE_CARLO_FILE_PATH "config/monte_carlo_events/focal/pixel_event_tree_pythia_MB_r4cm.root"
#define DEFAULT_FOCAL_STAVES_PER_QUADRANT "3"

#define DEFAULT_EVENT_RANDOM_HIT_GENERATION "true"
#define DEFAULT_EVENT_RANDOM_CLUSTER_GENERATION "false"
#define DEFAULT_EVENT_RANDOM_CLUSTER_SIZE_MEAN "4"
#define DEFAULT_EVENT_RANDOM_CLUSTER_SIZE_STDDEV "2"
#define DEFAULT_EVENT_MONTE_CARLO_FILE_TYPE "xml"
#define DEFAULT_EVENT_QED_NOISE_PATH "config/monte_carlo_events/QED"
#define DEFAULT_EVENT_QED_NOISE_INPUT "false"
#define DEFAULT_EVENT_QED_NOISE_FEED_RATE_NS "250"
#define DEFAULT_EVENT_QED_NOISE_EVENT_RATE_NS "20000"
#define DEFAULT_EVENT_TRIGGER_DELAY_NS "1000"
#define DEFAULT_EVENT_TRIGGER_FILTER_TIME_NS "10000"
#define DEFAULT_EVENT_TRIGGER_FILTER_ENABLE "true"
#define DEFAULT_EVENT_STROBE_ACTIVE_LENGTH_NS "100"
#define DEFAULT_EVENT_STROBE_INACTIVE_LENGTH_NS "100"
#define DEFAULT_EVENT_AVERAGE_EVENT_RATE_NS "2500"

QSettings *getSimSettings(const char *fileName = "config/settings.txt");
void setDefaultSimSettings(QSettings *readoutSimSettings);


#endif
///@}
